WalletWasabi/WabiSabi/Client/DependencyGraphResolver.cs

```diff
@@ -0,0 +1,201 @@
+
+			var amountEdges = Graph.Inputs.SelectMany(node => Graph.OutEdges(node, CredentialType.Amount));
+			var vsizeEdges = Graph.Inputs.SelectMany(node => Graph.OutEdges(node, CredentialType.Vsize));
+
+			// Check if all tasks were finished, otherwise Task.Result will block.
+			if (!amountEdges.Concat(vsizeEdges).All(edge => DependencyTasks[edge].Task.IsCompletedSuccessfully))
+			{
+				throw new InvalidOperationException("All Input nodes' outedges should be completed.");
+			}
+		}
+
+		public async Task StartReissuancesAsync(IEnumerable<AliceClient> aliceClients, BobClient bobClient, CancellationToken cancellationToken)
+		{
+			var aliceNodePairs = PairAliceClientAndRequestNodes(aliceClients, Graph);
+
+			// Build tasks and link them together.
+			List<SmartRequestNode> smartRequestNodes = new();
+			List<Task> alltask = new();
+
+			using CancellationTokenSource ctsOnError = new();
+			using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ctsOnError.Token);
+
+			foreach (var node in Graph.Reissuances)
+			{
+				var inputAmountEdgeTasks = Graph.InEdges(node, CredentialType.Amount).Select(edge => DependencyTasks[edge].Task);
+				var inputVsizeEdgeTasks = Graph.InEdges(node, CredentialType.Vsize).Select(edge => DependencyTasks[edge].Task);
+
+				var outputAmountEdgeTaskCompSources = Graph.OutEdges(node, CredentialType.Amount).Select(edge => DependencyTasks[edge]);
+				var outputVsizeEdgeTaskCompSources = Graph.OutEdges(node, CredentialType.Vsize).Select(edge => DependencyTasks[edge]);
+
+				var requestedAmounts = Graph.OutEdges(node, CredentialType.Amount).Select(edge => (long)edge.Value);
+				var requestedVSizes = Graph.OutEdges(node, CredentialType.Vsize).Select(edge => (long)edge.Value);
+
+				SmartRequestNode smartRequestNode = new(
+					inputAmountEdgeTasks,
+					inputVsizeEdgeTasks,
+					outputAmountEdgeTaskCompSources,
+					outputVsizeEdgeTaskCompSources,
+					ZeroAmountCredentialPool,
+					ZeroVsizeCredentialPool);
+
+				var task = smartRequestNode
+					.StartReissuanceAsync(bobClient, requestedAmounts, requestedVSizes, linkedCts.Token)
+					.ContinueWith((t) =>
+				{
+					if (t.IsFaulted && t.Exception is { } exception)
+					{
+						// If one task is failing, cancel all the tasks and throw.
+						ctsOnError.Cancel();
+						throw exception;
+					}
+				}, linkedCts.Token);
+				alltask.Add(task);
+			}
+
+			await Task.WhenAll(alltask).ConfigureAwait(false);
+
+			var amountEdges = Graph.Outputs.SelectMany(node => Graph.InEdges(node, CredentialType.Amount));
+			var vsizeEdges = Graph.Outputs.SelectMany(node => Graph.InEdges(node, CredentialType.Vsize));
+
+			// Check if all tasks were finished, otherwise Task.Result will block.
+			if (!amountEdges.Concat(vsizeEdges).All(edge => DependencyTasks[edge].Task.IsCompletedSuccessfully))
+			{
+				throw new InvalidOperationException("All Output nodes' inedges should be completed.");
```

Comment:
Same suggestion that in previous log entry.

Reply:
ACK